app/models/bulk_operation.rb

```diff
@@ -1,34 +1,54 @@
 class BulkOperation < ApplicationRecord
   belongs_to :admin
+  belongs_to :ran_by_admin, class_name: "Admin", optional: true
   has_one_attached :file
 
   validate :file_valid
 
+  before_save :update_row_count
+
   scope :not_started, -> { where(started_at: nil) }
 
+  HEADERS = false
+
   def started?
     started_at.present?
   end
 
+  def ids_to_update
+    file.open { CSV.read(_1, headers: false).flatten }
+  end
+
 private
 
   def file_valid
     return unless file.attached?
 
-    errors.add(:file, :empty) unless file.blob.byte_size.positive?
-    if attachment_changes["file"]
-      check_format(attachment_changes["file"].attachable.read)
+    if file.blob.byte_size.positive?
+      check_format if attachment_changes["file"]
+    else
+      errors.add(:file, :empty)
     end
   end
 
-  def check_format(string)
-    CSV.parse(string) do |row|
-      if row.size > 1
-        errors.add(:file, :invalid)
-        break
-      end
-    end
+  def attached_file
+    attachment_changes["file"].attachable
```

Comment:
Is there a scenario where `attachment_changes["file"]` might be `nil` ? If so can this be changed to `&.attachable`

Reply:
I notice lower down in `row_count` it is checking for whether `attachment_changes["file"]` is nil?